src/query/phrase_query/phrase_scorer.rs

```diff
@@ -104,38 +105,104 @@ fn intersection_count(left: &[u32], right: &[u32]) -> usize {
 ///
 /// Returns the length of the intersection
 fn intersection(left: &mut [u32], right: &[u32]) -> usize {
-    let mut left_i = 0;
-    let mut right_i = 0;
+    let mut left_index = 0;
+    let mut right_index = 0;
     let mut count = 0;
     let left_len = left.len();
     let right_len = right.len();
-    while left_i < left_len && right_i < right_len {
-        let left_val = left[left_i];
-        let right_val = right[right_i];
+    while left_index < left_len && right_index < right_len {
+        let left_val = left[left_index];
+        let right_val = right[right_index];
         match left_val.cmp(&right_val) {
             Ordering::Less => {
-                left_i += 1;
+                left_index += 1;
             }
             Ordering::Equal => {
                 left[count] = left_val;
                 count += 1;
-                left_i += 1;
-                right_i += 1;
+                left_index += 1;
+                right_index += 1;
             }
             Ordering::Greater => {
-                right_i += 1;
+                right_index += 1;
             }
         }
     }
     count
 }
 
+/// Intersect twos sorted arrays `left` and `right` and outputs the
+/// resulting array in left.
+///
+/// Condition for match is that the value stored in left is less than
+/// the value in right and that the distance to the previous token is lte to the slop.
+///
+/// Returns the length of the intersection
+fn intersection_with_distance(
+    left: &mut [u32],
+    right: &mut [u32],
+    max_distance_to_begin: u32,
+) -> usize {
+    let mut left_index = 0;
+    let mut right_index = 0;
+    let mut count = 0;
+    let left_len = left.len();
+    let right_len = right.len();
+    // Is the current last value guaranteed to be the final value.
+    let mut is_temporary = false;
+    while left_index < left_len && right_index < right_len {
+        let left_val = left[left_index];
+        let right_val = right[right_index];
+        match left_val.cmp(&right_val) {
+            Ordering::Less => {
+                if right_val - left[left_index] <= max_distance_to_begin {
+                    if is_temporary {
```

Comment:
Once the condition is switch to the ternary choice above, the is_tempory could be avoided with an inner loop.